```
======================================================================
ERROR: test_basic_completed_by_user (corehq.apps.reports.tests.test_esaccessors.TestFormESAccessorsSQL)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/corehq/apps/reports/tests/test_esaccessors.py", line 78, in test_basic_completed_by_user
    self._send_form_to_es(completion_time=datetime(2013, 7, 2))
  File "/mnt/corehq/apps/reports/tests/test_esaccessors.py", line 183, in _send_form_to_es
    form_pair = make_es_ready_form(metadata, is_db_test=True)
  File "/mnt/corehq/util/test_utils.py", line 309, in make_es_ready_form
    wrapped_form = get_form_ready_to_save(metadata, is_db_test=is_db_test)
  File "/mnt/corehq/util/test_utils.py", line 297, in get_form_ready_to_save
    wrapped_form = process_xform_xml(metadata.domain, form_xml).submitted_form
  File "/mnt/corehq/form_processor/parsers/form.py", line 76, in process_xform_xml
    return _create_new_xform(domain, instance, attachments=attachments)
  File "/mnt/corehq/form_processor/parsers/form.py", line 117, in _create_new_xform
    if interface.is_duplicate(xform.form_id):
  File "/mnt/corehq/form_processor/interfaces/processor.py", line 103, in is_duplicate
    return self.processor.is_duplicate(xform_id, domain=domain)
  File "/mnt/corehq/form_processor/backends/sql/processor.py", line 51, in is_duplicate
    return FormAccessorSQL.form_with_id_exists(xform_id, domain=domain)
  File "/mnt/corehq/form_processor/backends/sql/dbaccessors.py", line 115, in form_with_id_exists
    cursor.execute('SELECT * FROM check_form_exists(%s, %s)', [form_id, domain])
  File "/vendor/lib/python2.7/site-packages/django/db/backends/utils.py", line 65, in execute
    return self.cursor.execute(sql, params)
  File "/vendor/lib/python2.7/site-packages/django/db/utils.py", line 94, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/vendor/lib/python2.7/site-packages/django/db/backends/utils.py", line 65, in execute
    return self.cursor.execute(sql, params)
ProgrammingError: function check_form_exists(unknown, unknown) does not exist
LINE 1: SELECT * FROM check_form_exists('6169ca92c15949fa84e73546f7a...
                      ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
```

--- a/corehq/apps/reports/tests/test_esaccessors.py
+++ b/corehq/apps/reports/tests/test_esaccessors.py
@@ -180,7 +180,7 @@ class TestFormESAccessorsSQL(TestFormESAccessors):
             time_end=completion_time or datetime.utcnow(),
             received_on=received_on or datetime.utcnow(),
         )
-        form_pair = make_es_ready_form(metadata, is_db_test=True)
+        form_pair = make_es_ready_form(metadata)
         change = Change(
             id=form_pair.json_form['form_id'],
             sequence_id='123',

--- a/corehq/form_processor/models.py
+++ b/corehq/form_processor/models.py
@@ -234,7 +234,9 @@ class XFormInstanceSQL(DisabledDbMixin, models.Model, RedisLockableMixIn, Attach
     @property
     def history(self):
         from corehq.form_processor.backends.sql.dbaccessors import FormAccessorSQL
-        return FormAccessorSQL.get_form_operations(self.form_id)
+        operations = FormAccessorSQL.get_form_operations(self.form_id) if self.is_saved() else []
+        operations += self.get_tracked_models_to_create(XFormOperationSQL)
+        return operations
 
     @property
     def metadata(self):

--- a/corehq/form_processor/models.py
+++ b/corehq/form_processor/models.py
@@ -21,6 +21,7 @@ from corehq.form_processor.track_related import TrackRelatedChanges
 from corehq.sql_db.routers import db_for_read_write
 from dimagi.utils.couch import RedisLockableMixIn
 from dimagi.utils.couch.safe_index import safe_index
+from dimagi.utils.couch.undo import DELETED_SUFFIX
 from dimagi.utils.decorators.memoized import memoized
 from dimagi.ext import jsonobject
 from couchforms.signals import xform_archived, xform_unarchived
@@ -449,6 +450,13 @@ class CommCareCaseSQL(DisabledDbMixin, models.Model, RedisLockableMixIn,
     case_json = JSONField(lazy=True, default=dict)
 
     @property
+    def doc_type(self):
+        dt = 'CommCareCase'
+        if self.is_deleted:
+            dt += DELETED_SUFFIX
+        return dt
+
+    @property
     def xform_ids(self):
         from corehq.form_processor.backends.sql.dbaccessors import CaseAccessorSQL
         return CaseAccessorSQL.get_case_xform_ids(self.case_id)
@@ -536,6 +544,10 @@ class CommCareCaseSQL(DisabledDbMixin, models.Model, RedisLockableMixIn,
         transactions = CaseAccessorSQL.get_transactions(self.case_id) if self.is_saved() else []
         transactions += self.get_tracked_models_to_create(CaseTransaction)
         return transactions
+
+    @property
+    def non_revoked_transactions(self):
+        return [t for t in self.transactions if not t.revoked]
 
     @property
     @memoized

--- a/corehq/form_processor/serializers.py
+++ b/corehq/form_processor/serializers.py
@@ -28,17 +28,21 @@ class CommCareCaseIndexSQLSerializer(serializers.ModelSerializer):
         model = CommCareCaseIndexSQL
 
 
-class CaseTransactionSerializer(serializers.ModelSerializer):
-    details = serializers.DictField()
+class CaseTransactionActionSerializer(serializers.ModelSerializer):
+    xform_id = serializers.CharField(source='form_id')
+    date = serializers.CharField(source='server_date')
 
     class Meta:
         model = CaseTransaction
-        exclude = ('id', 'case')
+        fields = ('xform_id', 'server_date', 'date', 'sync_log_id')
 
 
 class CommCareCaseSQLSerializer(serializers.ModelSerializer):
+    _id = serializers.CharField(source='case_id')
+    doc_type = serializers.CharField()
+    user_id = serializers.CharField(source='modified_by')
     indices = CommCareCaseIndexSQLSerializer(many=True, read_only=True)
-    transactions = CaseTransactionSerializer(many=True, read_only=True)
+    actions = CaseTransactionActionSerializer(many=True, read_only=True, source='non_revoked_transactions')
 
     class Meta:
         model = CommCareCaseSQL

--- a/corehq/pillows/case.py
+++ b/corehq/pillows/case.py
@@ -72,28 +72,6 @@ def transform_case_for_elasticsearch(doc_dict):
     return doc_ret
 
 
-def transform_case_transactions_to_actions(transactions):
-    def _action(trans):
-        return {
-            'server_date': trans['server_date'],
-            'date': trans['server_date'],
-            'xform_id': trans['form_id'],
-            'sync_log_id': trans['sync_log_id'],
-        }
-    return [_action(trans) for trans in transactions if not trans['revoked']]
-
-
-def prepare_sql_case_json_for_elasticsearch(sql_case_json):
-    prepped_case = transform_case_for_elasticsearch(sql_case_json)
-    # todo: these are required for consistency with couch representation, figure out how best to deal with it
-    prepped_case['doc_type'] = 'CommCareCase'
-    prepped_case['_id'] = prepped_case['case_id']
-    prepped_case['user_id'] = prepped_case['modified_by']
-    transactions = prepped_case.pop('transactions')
-    prepped_case['actions'] = transform_case_transactions_to_actions(transactions)
-    return prepped_case
-
-
 def get_sql_case_to_elasticsearch_pillow():
     checkpoint = PillowCheckpoint(
         get_django_checkpoint_store(),
@@ -102,7 +80,7 @@ def get_sql_case_to_elasticsearch_pillow():
     case_processor = ElasticProcessor(
         elasticseach=get_es_new(),
         index_meta=ElasticsearchIndexMeta(index=CASE_INDEX, type=CASE_ES_TYPE),
-        doc_prep_fn=prepare_sql_case_json_for_elasticsearch
+        doc_prep_fn=transform_case_for_elasticsearch
     )
     return ConstructedPillow(
         name='SqlCaseToElasticsearchPillow',

--- a/corehq/util/test_utils.py
+++ b/corehq/util/test_utils.py
@@ -287,6 +287,7 @@ def get_form_ready_to_save(metadata, is_db_test=False):
     from corehq.form_processor.parsers.form import process_xform_xml
     from corehq.form_processor.utils import get_simple_form_xml, convert_xform_to_json
     from corehq.form_processor.interfaces.processor import FormProcessorInterface
+    from corehq.form_processor.models import Attachment
 
     assert metadata is not None
     metadata.domain = metadata.domain or uuid.uuid4().hex
@@ -296,9 +297,13 @@ def get_form_ready_to_save(metadata, is_db_test=False):
     if is_db_test:
         wrapped_form = process_xform_xml(metadata.domain, form_xml).submitted_form
     else:
+        interface = FormProcessorInterface(domain=metadata.domain)
         form_json = convert_xform_to_json(form_xml)
-        wrapped_form = FormProcessorInterface(domain=metadata.domain).new_xform(form_json)
+        wrapped_form = interface.new_xform(form_json)
         wrapped_form.domain = metadata.domain
+        interface.store_attachments(wrapped_form, [
+            Attachment(name='form.xml', raw_content=form_xml, content_type='text/xml')
+        ])
     wrapped_form.received_on = metadata.received_on
     return wrapped_form
 

